_posts/2020-01-31-Everything-you-need-to-know-about-ISP.md

```diff
@@ -0,0 +1,166 @@
+image:
+  auto: 0018-cogs
+tags: ["Interface Segregation Principle", "ISP", "SOLID", "Principles of software development"]
+---
+
+
+>Clients should not be forced to depend upon interfaces that they do not use." — Robert Martin, ISP paper [Principles of OOD ](https://web.archive.org/web/20150905081110/http://www.objectmentor.com/resources/articles/isp.pdf)
+
+Abstraction is the heart of object-oriented design which allows the client to be unconcerned with the implementation details of functionality. In Java, abstraction is achieved through abstract classes and Interface. This article explains the idea of the Interface Segregation Principle which is one of  the [SOLID](https://en.wikipedia.org/wiki/SOLID) principles
+
+{% include github-project.html url="https://github.com/mukul-s/code-examples/tree/master/craft/interface%20segregation%20principle" %}
+
+## What is an Interface?
+An Interface is a set of abstractions which implementing class must follow. Basically, you define the behaviour but not implement them.
+For e.g. You can define them in Java as below:
+```java
+interface Dog{
+  void bark();
+}
+
+class Poodle implements dog{
+  // Poodle class needs to implement dog otherwise we will get a compilation error
+  public void bark(){
+    
+  }
+}
+```
+
+## What is the Interface Segregation Principle (ISP)
+ISP states that **A Client should not be exposed to methods that they don't need**. Declaring methods in an interface which it doesn't need pollutes the interface and leads to the bulky interface also called a fat interface.
+
+## Why you need ISP in your code?
+Lets taken an example to understand this - Below is code for **Burgerpoint** where a customer can order burger, fries or a combo.
+``` java 
+/**
+ * IOrder.java
+ **/
+public interface IOrder {
+    void orderBurger(int quantity);
+    void orderFries(int fries);
+}
+```
+Since a customer can order fries or burger both, we decided to put both of them in a single interface. Now to support a burger only order, we have to throw an exception in **orderFries** method.
+```java
+/**
+ * BurgerOrder.java
+ **/
+public class BurgerOrder implements IOrder {
+    @Override
+    public void orderBurger(int quantity) {
+
+    }
+
+    @Override
+    public void orderFries(int fries) {
+        throw new UnsupportedOperationException("No fries in Burger only order");
+    }
+}
+
+// Similarly for FriesOrder.java
+```
+Similarly, for fries only order you also need to throw an exception in **orderBurger** method. This is not the only downside of this design, these classes will also have unwanted side effects, whenever you make changes to your abstraction. Let's say you decided to accept an order of fries in multiple units such as pounds or grams. In that case we most likely we have to add a parameter for the unit in orderFries method. This change will also affect the **BurgerOrder.java** even when it doesn't really support the ordering of fries. 
```

Comment:
Noted, Although the title is still "Everything **you** need to know.". I have left it as it is since not mentioned in comments, let me know if this needs to be changed too